### PR TITLE
create vendor bundle with hand-picked libraries

### PR DESCRIPTION
**Pull request by davidsbailey**
### Background

One problem with our current webpack config is that shared modules are not pulled out of plc entry points. Because using count-based approaches is [tricky](https://github.com/code-dot-org/code-dot-org/blob/e9995bd5fa3a470fa0cca6ec8c0094487dcad533/apps/Gruntfile.js#L790-L811), I am hand-picking the list of packages included here based on inspection of which packages are used by most entry points as well as which packages are aligned with our current best practices (e.g. react). 

According to many sources on the internet, the strategy of bundling npm packages separately from application code should improve our caching performance because vendors.js is likely to remain cached (by Cloudfront, firewalls, browsers, etc.) for days or weeks, compared with our application code which updates daily.

### Description

Extract a hand-picked list of npm packages from most dashboard and pegasus entry points. This does not include special entry points like applab-api.js and embedVideo.js which need to be entirely self-contained. 

### results

* overall bundle size reduced by **11MB** (from 49.4 to 38.4)
* many PLC entry points now include 700KB less JS
* code-studio-common.js decreased in size by 700KB (the size of vendors.js), therefore JS size does not increase for any dashboard page.
* theoretically improves cache performance, because the 700KB moved to vendors.js can be cached for days or weeks whereas code-studio-common.js is typically modified daily.

### before
Screen Shot 2019-08-22 at 1 27 50 PM

### after
Screen Shot 2019-08-22 at 1 27 59 PM

### future work

there is some code still shared across many PLC entry points (e.g. react-bootstrap) which could benefit from being extracted into some kind of `plc-common.js` bundle. I didn't include this in vendors.js because react-bootstrap is not used widely outside of PLC code. 

Ideally, we'd be able to do something like this:
* app-specific code extracted into common.js
* plc-common code extracted into plc-common.js
* any code common to common.js and plc-common.js further extracted into code-studio-common.js
currently we use this strategy to avoid duplication between common.js and code-studio-common.js, however it is not clear whether the [trick](https://github.com/code-dot-org/code-dot-org/blob/e9995bd5fa3a470fa0cca6ec8c0094487dcad533/apps/Gruntfile.js#L790-L811) mentioned earlier will scale as the number of shared bundles increases.

I have a future work item to explore a more dynamic approach to bundle splitting which does not require our configuration to have specific knowledge of our application code, which is part of why I'm choosing this as a stopping point for now.

### CODE REVIEW

File: apps/Gruntfile.js [hunk 1]
```diff
@@ -819,6 +820,38 @@ describe('entry tests', () => {
                 return chunkNames.includes(chunk.name);
               },
               priority: 20
+            },
+            vendors: {
+              name: 'vendors',
+              priority: 30,
```
Comment (joshlory) [inline apps/Gruntfile.js:829]: What does this do?

**Comment by joshlory** [comment]
This is cool! Can you share a before and after snapshot of the `build:analyze` report?

File: apps/Gruntfile.js [hunk 2]
```diff
@@ -819,6 +820,38 @@ describe('entry tests', () => {
...
+              chunks: chunk => {
+                // all 'initial' chunks except otherEntries
+                const chunkNames = _.concat(
+                  _.keys(codeStudioEntries),
+                  _.keys(appsEntries),
+                  _.keys(pegasusEntries),
+                  _.keys(professionalDevelopmentEntries),
+                  _.keys(internalEntries)
```
Comment (joshlory) [inline apps/Gruntfile.js]: Nit: why not `Object.keys`?

File: apps/Gruntfile.js (hunk 1 above)
Comment (davidsbailey) [reply apps/Gruntfile.js:829]: we discussed offline briefly. more details are here: https://webpack.js.org/plugins/split-chunks-plugin/#splitchunkscachegroupscachegrouppriority

File: dashboard/app/views/admin_styleguide/show.html.haml [hunk 3]
```diff
@@ -1,7 +0,0 @@
-#styleguide-app
-%script{src: minifiable_asset_path('js/marked/marked.js')}
-%script{src: minifiable_asset_path('js/essential.js')}
-%script{src: minifiable_asset_path('js/code-studio-common.js')}
-%script{src: minifiable_asset_path('js/common.js')}
-%script{src: asset_path("js/#{js_locale}/common_locale.js")}
-%script{src: minifiable_asset_path('js/styleguide.js')}
```
Comment (joshlory) [inline dashboard/app/views/admin_styleguide/show.html.haml:7]: Removed because it already inherits from the application template?

**Review by joshlory — APPROVED** [review]

**Comment by joshlory** [comment]
Caveat: I haven't tested this locally yet. Lmk if you'd like me to!

File: dashboard/app/views/admin_styleguide/show.html.haml (hunk 3 above)
Comment (davidsbailey) [reply dashboard/app/views/admin_styleguide/show.html.haml:7]: Removed because it unused and would require updating otherwise.

**Comment by davidsbailey** [comment]
I don't need you to test this locally unless you want to but thanks for the careful review!

**Pull request merged**